timemachine/fe/dummy.py

```diff
@@ -0,0 +1,631 @@
+                    # this looks weird at first, but is actually valid
+                    # consider:
+                    # i j
+                    # |/
+                    # a-b
+                    # we can freely rotate i-j independently of a-b
+                    # in both of the following cases.
+                    allowed_ixns.add((i, a, b, j))
+                    allowed_ixns.add((a, b, i, j))
+
+                    # disallowed examples would be (i,a,j,b)
+
+            for (i,) in nc1:
+                allowed_ixns.add((i, a, b))
+                # consider
+                # i
+                # |
+                # a-b
+                # we're still allowed to have an angle term (a,b,i) without
+                # affecting the factorizability
+                allowed_ixns.add((a, b, i))
+
+                # disallowed examples would be (a,i,b)
+
+            if len(ag) > 2:
+                c = ag[2]
+                for (i,) in nc1:
+                    allowed_ixns.add((i, a, b, c))
+                    # this is also allowed (and encountered in a trefoil)
+                    # consider:
+                    #    i
+                    #   /
+                    #  / b--c
+                    # a-/
+                    allowed_ixns.add((a, b, c, i))
+
+    return make_bond_set(allowed_ixns)
+
+
+def find_ags_for_dg(mol, core, dg):
+    """
+    Find all possible anchor groups for a given dummy group.
+
+    Parameters
+    ----------
+    mol: Chem.Mol
+        input molecule
+
+    core: list of int
+        core indices
+
+    dg: iterable of int
+        dummy group
+
+    Returns
+    -------
+    list of list of int of up to size 3
+        Returns all allowed anchor groups for a given dummy group.
+
+    """
+    anchors = []
+    for dummy_atom in dg:
+        anchors.extend(identify_root_anchors(mol, core, dummy_atom))
+    anchors = set(anchors)
+
+    anchor_groups = []
+    for anchor in anchors:
+        ag0, ag1, ag2 = identify_anchor_groups(mol, core, anchor)
+        anchor_groups.extend(ag0)
+        anchor_groups.extend(ag1)
+        anchor_groups.extend(ag2)
+
+    return anchor_groups
+
+
+def ordered_tuple(ixn):
+    if ixn[0] > ixn[-1]:
+        return tuple(ixn[::-1])
+    else:
+        return tuple(ixn)
+
+
+def make_bond_set(old_set):
+    """
+    Bond set has the requirement that indices are symmetrically
+    reversible. i.e. ij = ji, ijk = kji, ijkl = lkji
+    """
+    new_set = set()
+    for idxs in old_set:
+        new_set.add(ordered_tuple(idxs))
+    return new_set
+
+
+def flag_bonds(mol, core, bond_idxs):
+    """
+    Flags bonds based on minimizing the number of terms we have to
+    turn off.
+
+    Parameters
+    ----------
+    mol: Chem.Mol
+        rdkit molecule
+
+    core: list of int
+        idxs of core atoms
+
+    bond_idxs: list of list of int
+        list of 2-tuple, 3-tuple, 4-tuple
+
+    Returns
+    -------
+        boolean flags, 1: keep, 0: remove
```

Comment:
nit: is this always a list of the same length as `bond_idxs`?

Reply:
We now only support strict mode - so the first comment is less relevant.